// @bitrix24/b24jssdk/dist/esm/core/interaction/batch/processing/interface-strategy.mjs
/**
 * @package @bitrix24/b24jssdk
 * @version 1.0.4
 * @copyright (c) 2026 Bitrix24
 * @license MIT
 * @see https://github.com/bitrix24/b24jssdk
 * @see https://bitrix24.github.io/b24jssdk/
 */
import { AjaxError } from '../../../http/ajax-error.mjs';

var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
class AbstractProcessing {
  static {
    __name(this, "AbstractProcessing");
  }
  // endregion ////
  // region Tools ////
  _getBatchResultByIndex(data, index) {
    if (!data) return void 0;
    if (Array.isArray(data)) {
      return data[index];
    }
    return data[index];
  }
  _createErrorFromAjaxResult(ajaxResult) {
    if (ajaxResult.hasError("base-error")) {
      return ajaxResult.errors.get("base-error");
    }
    return new AjaxError({
      code: "JSSDK_BATCH_SUB_ERROR",
      description: ajaxResult.getErrorMessages().join("; "),
      status: ajaxResult.getStatus(),
      requestInfo: { ...ajaxResult.getQuery() },
      originalError: ajaxResult.getErrors().next().value
    });
  }
  // endregion ////
}

export { AbstractProcessing };
